### PR TITLE
Added configurable timeout to connection string

### PR DESCRIPTION
**Pull request by sergey-ushakov**
* RedisURI.java, RedisURITest.java - add timeout configuration to Redis connection string

### CODE REVIEW

File: src/main/java/com/lambdaworks/redis/RedisURI.java [hunk 1]
```diff
@@ -170,10 +172,67 @@ public static RedisURI create(URI uri) {
...
 
     }
 
+    private static void parseTimeout(Builder builder, String queryParam) {
+        int index = queryParam.indexOf('=');
+        if (index < 0) {
+            return;
+        }
+
+        String timeoutString = queryParam.substring(index + 1);
```
Comment (mp911de) [inline src/main/java/com/lambdaworks/redis/RedisURI.java]: How about decoding the value part using `URLDecoder.decode`?

File: src/main/java/com/lambdaworks/redis/RedisURI.java [hunk 2]
```diff
@@ -170,10 +172,67 @@ public static RedisURI create(URI uri) {
...
+        }
+
+        if (numbersEnd == 0) {
+            if (timeoutString.startsWith("-")) {
+                builder.withTimeout(0, TimeUnit.MILLISECONDS);
+            } else {
+                // no-op, leave defaults
+            }
+        } else {
+            long timeoutValue = Long.parseLong(timeoutString, 0, numbersEnd, 10);
```
Comment (mp911de) [inline src/main/java/com/lambdaworks/redis/RedisURI.java]: This line causes compile errors, see https://travis-ci.org/mp911de/lettuce/builds/90676460

File: src/test/java/com/lambdaworks/redis/RedisURITest.java [hunk 3]
```diff
@@ -85,4 +86,34 @@ public void socketEqualsTest() throws Exception {
         assertThat(redisURI3).isNotEqualTo(redisURI2);
         assertThat(redisURI3.hashCode()).isNotEqualTo(redisURI2.hashCode());
     }
+
+    @Test public void timeoutParsingTest() {
+        checkUriTimeout("redis://auth@localhost:1234/5?timeout=5000", 5000, TimeUnit.MILLISECONDS);
```
Comment (mp911de) [inline src/test/java/com/lambdaworks/redis/RedisURITest.java]: Cool, great test coverage.

File: src/main/java/com/lambdaworks/redis/RedisURI.java [hunk 4]
```diff
@@ -170,10 +172,67 @@ public static RedisURI create(URI uri) {
...
+            builder.withTimeout(timeoutValue, TimeUnit.MILLISECONDS);
+
+            String suffix = timeoutString.substring(numbersEnd);
+            TimeUnit timeoutUnit;
```
Comment (mp911de) [inline src/main/java/com/lambdaworks/redis/RedisURI.java]: How about putting the units in a lookup map?

**Comment by mp911de** [comment]
Thanks for the PR. I was tempted all the time to introduce options within the query string and you made the first step :+1: 
I added some comments to the code. The PR runs currently into build errors. Can you fix those and adjust the code formatting within the test?

**Comment by sergey-ushakov** [comment]
Query string seems to be already decoded, added test for it. Also added lookup map.

**Comment by mp911de** [comment]
Great, thank you.

**Pull request merged**